R-package/R/lgb.Booster.R

```diff
@@ -486,6 +486,34 @@ Booster <- R6::R6Class(
         start_iteration <- 0L
       }
 
+      # possibly override keyword arguments with parameters
+      #
+      # NOTE: this length() check minimizes the latency introduced by these checks,
+      #       for the common case where params is empty
+      #
+      # NOTE: doing this here instead of in Predictor$predict() to keep
+      #       Predictor$predict() as fast as possible
```

Comment:
Why was this needed?

Reply:
Like I mentioned in https://github.com/microsoft/LightGBM/pull/4909#pullrequestreview-842363693 and https://github.com/microsoft/LightGBM/pull/4977#pullrequestreview-894601022, I think we should be overly-protective of `predict()` functions / methods, as those may be used in latency-sensitive applications like serving a model as a REST API.

Each call to `lgb.check.wrapper_param()` results in a call to get all aliases from the C++ side and deserialize them from JSON into an R list: 

https://github.com/microsoft/LightGBM/blob/b462d0a40f6978c0a66a47af4522c2ae12532316/R-package/R/utils.R#L193

https://github.com/microsoft/LightGBM/blob/b462d0a40f6978c0a66a47af4522c2ae12532316/R-package/R/aliases.R#L40-L45

Since I expect that the use of additional `params` in `predict()` is not common, I added this `if length(` statement here to avoid imposing the overhead of those 3 `lgb.check.wrapper_param()` calls on existing user code if it wasn't necessary.

Reply:
Hmm, thanks for the detailed as always explanation! Are you aware of any `Predictor` usage examples in a third-party packages?

Is it possible to refactor `.PARAMETER_ALIASES()` in a way it will call cpp function only once for the session. Refer to the Python-package implementation as an example:
https://github.com/microsoft/LightGBM/blob/b462d0a40f6978c0a66a47af4522c2ae12532316/python-package/lightgbm/basic.py#L340

Reply:
> Are you aware of any Predictor usage examples in a third-party packages?

I'm not aware of any such usage. It is *technically* possible for user code to get a `Predictor` using the public API of the R package.

You'd have to use `lgb.train()` or `lightgbm()` to get a `Booster`, and then `Booster$to_predictor()` to get a `Predictor`.

https://github.com/microsoft/LightGBM/blob/b462d0a40f6978c0a66a47af4522c2ae12532316/R-package/R/lgb.Booster.R#L509-L511

But I would be surprised if many users do that, so for all intents and purposes `Predictor` is private today.

> Is it possible to refactor `.PARAMETER_ALIASES()` in a way it will call cpp function only once for the session

Yes, it is. Briefly, I can think of at least 3 options:

1. Have an environment in the R package's namespace, and have calls to `.PARAMETER_ALIASES()` check that environment for a list like `parameter_aliases`, and only re-populate it if its missing
    - this could also possibly be loaded at the time the package is loaded, using the special `.onLoad()` hook (see ["load hooks" in Writing R extensions](https://cran.r-project.org/doc/manuals/r-release/R-exts.html#Load-hooks))
2. Use the `{memoise}` package ([link](https://cran.r-project.org/web/packages/memoise/index.html)), which has functionality similar to `functools.lru_cache()` in Python.

This issue is larger than just this PR, since `.PARAMETER_ALIASES()` is used many times in the R package. I'll write up a separate issue describing this and link it here.

Reply:
> I'll write up a separate issue

Ok actually it was easy enough to just put my preferred proposal directly into a pull request, so I did that instead of creating an issue.

I've opened #5141.

Note that even if that PR was merged, I'd still recommend keeping this `if (length(params) > 0L)` check, to avoid running the other R-side for loops and checks in `lgb.check.wrapper_param()` on calls to `predict()`.

https://github.com/microsoft/LightGBM/blob/b462d0a40f6978c0a66a47af4522c2ae12532316/R-package/R/utils.R#L193-L203

Reply:
> Note that even if that PR was merged, I'd still recommend keeping this `if (length(params) > 0L)` check, to avoid running the other R-side for loops and checks in `lgb.check.wrapper_param()` on calls to `predict()`.

I don't have anything against `if (length(params) > 0L)` check! 🙂 
I was just thinking about the reasons to keep it in `Booster` instead of `Predictor`.

Reply:
Ah ok!

I think it's preferable to keep it in the `Booster` for now, since `Predictor$predict()`  doesn't support taking `params`

https://github.com/microsoft/LightGBM/blob/9a964fef17c5e20587b5d4f3120011065ad1b08b/R-package/R/lgb.Predictor.R#L81-L87

and since it needs to pass boolean values (well, integer representations of them) for raw_score, leaf_index, and pred_contrib to the `LGBM_Predict*`  functions

https://github.com/microsoft/LightGBM/blob/9a964fef17c5e20587b5d4f3120011065ad1b08b/R-package/R/lgb.Predictor.R#L110-L122

Since these are all internal details, I think we can change it in the future without breaking any user code if there's a compelling reason to move this into the `Predictor`.

Reply:
> `Predictor$predict()`  doesn't support taking `params`

Yeah, it takes them in a constructor
https://github.com/microsoft/LightGBM/blob/f860282658d079f517cce7b614a1f08e6ccc6d3e/R-package/R/lgb.Predictor.R#L29
https://github.com/microsoft/LightGBM/blob/9a964fef17c5e20587b5d4f3120011065ad1b08b/python-package/lightgbm/basic.py#L707-L708

Reply:
Right, it takes them in the constructor and then immediately converts them to a string to be passed to the C++ side and stores that string.

https://github.com/microsoft/LightGBM/blob/0a4851f5fed79fc8815e97c23e92756ab77bef1c/R-package/R/lgb.Predictor.R#L30

Since `Predictor$predict()` accepts things like `rawscore` as keyword arguments, supporting passing something like `predict_raw_score` in `params` would require one of the following changes to `Predictor`:

* extract `rawscore`, `predleaf`, etc. in the constructor and store them as private attributes, before converting to string with `lgb.params2str()`
* accept a `params` list in `Predictor$predict()` instead of in the constructor
* accept a `params` list in `Predictor$predict()` and merge it with whatever `params` were passed into the constructor and stored at `private$params`

I think it'd be better to do this work up in `Booster` than to make any of those changes to `Predictor` to support this.

That would match the behavior of LightGBM's C API anyway...the root cause of #4670 and the reason for this PR in the first place is that functions like `LGBM_BoosterPredictForMat()` on the C++ side ignore parameters like `predict_raw_score` passed through `params`, and instead rely on `predict_type` being resolved outside and passed through a function argument.

https://github.com/microsoft/LightGBM/blob/0a4851f5fed79fc8815e97c23e92756ab77bef1c/R-package/R/lgb.Predictor.R#L159-L172

https://github.com/microsoft/LightGBM/blob/0a4851f5fed79fc8815e97c23e92756ab77bef1c/R-package/src/lightgbm_R.cpp#L839

https://github.com/microsoft/LightGBM/blob/0a4851f5fed79fc8815e97c23e92756ab77bef1c/R-package/src/lightgbm_R.cpp#L846-L848

https://github.com/microsoft/LightGBM/blob/0a4851f5fed79fc8815e97c23e92756ab77bef1c/src/c_api.cpp#L2133

I think parameters like `predict_raw_score` only exist on the C++ side for the CLI.

https://github.com/microsoft/LightGBM/blob/6b56a90cd1324a6dbac2afa0a352c9355b0dc3cf/src/application/application.cpp#L248-L251

Reply:
@StrikerRUS since you and @jmoralez have approved this PR and since in this thread we're only discussing the internal API of the package (so decisions are reversible without impacting users), I've updated this to latest `master` and am going to merge it once CI passes.

If you come back to this thread and want something to be changed, please let me know.

Reply:
Sure, fine with me!

Reply:
Thanks! This was my plan last night....then all CI started breaking, with the issues being discussed on #5169 😂